### ISSUE TITLE
[ENH]: figure clear=True should clear events to prevent duplicates

### ISSUE THREAD

**Issue by davidgilbertson** [matplotlib/matplotlib#27508]
### Problem

When I run this code in an interactive console
```py
from matplotlib import pyplot as plt

plt.rcParams["backend"] = "TKAgg"
plt.ion()
fig, ax = plt.subplots(num="My chart", clear=True)

def on_click(event):
    print("Hello")

fig.canvas.mpl_connect("button_press_event", on_click)
```

and then rerun the code in the same interactive session (either in PyCharm, VS Code, or a Jupyter notebook) the same figure is reused (good) but the connected events are never removed, so I get a buildup of duplicated events.

### Proposed solution

I've written a solution for this, in my own utilities file, I track what the built in events are, and disconnect any user-added events, like so:

```py
# Connection IDs present in a new figure.
_initial_cids = {}


def clear_events(fig=None):
    """
    Clears user-added event handlers so that code can be re-run without
    a build up of events.
    Should be called after figure creation, before adding events
    """
    fig = fig or plt.gcf()

    # Get all the CIDs in the callback registry
    cids = []
    for cb_dict in fig.canvas.callbacks.callbacks.values():
        cids.extend(cb_dict.keys())

    if _initial_cids.get(fig.number) is None:
        # These are the event CIDs for a new figure
        _initial_cids[fig.number] = cids
    else:
        # We must be re-running code for an existing figure
        # What CIDs have been added since the first time this was called?
        added_cids = set(cids).difference(_initial_cids[fig.number])

        # Disconnect those CIDs
        for cid in added_cids:
            fig.canvas.mpl_disconnect(cid)
```

Which is then used like so:

```py
from matplotlib import pyplot as plt

import my_utils

plt.rcParams["backend"] = "TKAgg"
plt.ion()
fig, ax = plt.subplots(num="My chart", clear=True)
my_utils.clear_events()  # <- clear any previously-added event handlers, below will re-add them

def on_click(event):
    print("Hello")

fig.canvas.mpl_connect("button_press_event", on_click)
```

I've been using this for a while and it works fine, which leaves me wondering a few things:
* Am I doing something dumb, do other people not have this same problem, is there a more obvious solution that I'm missing?
* If this is a sensible solution, could it be baked into Matplotlib?
* If so, could it be run by default when `clear=True` is passed during figure creation? Or is there a legitimate case where a user wants to use an existing figure, has passed `clear=True`, but expects the events to not clear.

**Comment by ksunden** [comment]
I think there are extremely valid reasons to desire current behavior, so I'm not totally convinced it _should_ change... 

First of all it is a question of what it means to "clear" a figure... the event handling is done on the `FigureCanvas`, not on the `Figure` object itself, and when cleared, mostly it means simply removing all artists from the figure, but keeping the current canvas (just it is now blank), including e.g. the canvas size and dpi.

To make things more complicated, `Figure.clear` takes a `keep_observers` argument (defaulting to `False`) which _does_ affect a _different_ `CallbackRegistry` object (one that lives on the `Figure` directly and is interacted with via `add_axobserver`, a method that does _not_ return a CID and does not provide any way of removing _other_ than `clear`, and only provides one signal "_axes_change_event", which is triggered on `sca` and `_remove_axes`. This sounds kind of like it could interact with the canvas callbacks, but it doesn't, and I think it is intended to be quite a bit narrower...

Personally I think it makes sense that canvas-local attributes (like the callbacks) are _not_ reset under this description of "clear".

Even in the case of the built-in viewer applicatioan, we use the callbacks to build the UI to interact with the figure. It is not unreasonable to think that it is _likely_ that some users are extending the UI with additional callbacks that they _don't_ want wiped away when clearing out the figure. These callbacks are added at various places in backend code/backend_bases/widgets, etch and not in a way that you could e.g. clear out all of them and then easily add back in the default set.

I don't particularly like having to keep track of "user added" event handlers... that seems more likely to cause problems than solve them to me (e.g. all of the cases where that cache should be invalidated). I also do not like the idea of relying on order of execution/CIDs starting at 0 to identify the "default" set as is in the example here... that seems fragile to me. (We could probably do _better_ implementing such tracking internally... but not confident we could do _precisely_ what we want...)

**Comment by davidgilbertson** [comment]
OK so it sounds like `clear=True` clearing events is out, that's fair enough.

May I pivot the feature request then to suggest adding a way to clear events (e.g. via a new `clear_events` parameter as a sibling to `clear`).

The only reason my code keeps track of user-added event handlers is to remain robust over Matplotlib versions (in case the next version has extra/fewer built in event handlers), because it's external to Matplotlib. If the behaviour were to be moved into Matplotlib, then obviously there could be a more direct way to remove them, as you say.

BTW there's no reliance on order of CIDs here, the logic would work if MPL switched to UUIDs or anything else for CIDs, it's just looking for CIDs that weren't there immediately after a new figure was created, not CIDs that are larger or anything like that.

**Comment by FilipDominec** [comment]
I agree that duplicate calls of event handlers should be adressed in some systematic manner. 

This is particularly important in an interactive data browser; I am developing and commonly using a frontend to matplotlib https://github.com/FilipDominec/nihilnovi/, but certainly many other projects have ran into the same issue. After some hours of data viewing, which involves replotting the figure, dozens of mouse click handlers can accumulate and be called upon every mouse click. 

The @davidgilbertson's proposal of `clear_events()` looks neat, or perhaps matplotlib could implement e.g. a `remove_previous` keyword for the existing function, that would do the same job prior to registering a new handler... 

```fig.canvas.mpl_connect("button_press_event", on_click, remove_previous=True)```

This seems to be also an intuitive and backward compatible solution.
